File: examples/python/counter_client.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Sui Move Counter Client - Python SDK Example
"""

import base64
from typing import Optional

from pysui.sui.sui_clients import SyncClient
from pysui.sui.sui_config import SuiConfig
from pysui.sui.sui_builders.move_call import MoveCallTransaction
from pysui.sui.sui_builders.get_object import GetObject
from pysui.sui.sui_types.address import SuiAddress

class CounterClient:
    """
    Client for interacting with the Counter contract on Sui blockchain.
    """
    
    def __init__(self, config_path: Optional[str] = None, package_id: str = None):
        """
        Initialize the counter client.
        
        Args:
            config_path: Path to Sui config file (defaults to ~/.sui/sui_config/client.yaml)
            package_id: ID of the deployed counter package
        """
        # Initialize Sui client configuration
        self.config = SuiConfig.from_config_file(config_path) if config_path else SuiConfig.default_config()
        self.client = SyncClient(self.config)
        self.package_id = package_id
        self.counter_module = "counter"
    
    def create_counter(self) -> str:
        """
        Create a new counter object.
        
        Returns:
            ID of the created counter object
        """
        tx = MoveCallTransaction(
            self.client, 
            signer=self.config.active_address,
            package_object_id=self.package_id,
            module=self.counter_module,
            function="create_and_transfer",
            arguments=[],
            type_arguments=[]
        )
        
        result = tx.execute()
        
        if not result.is_ok():
            raise Exception(f"Failed to create counter: {result.result_string}")
            
        # Extract the created object ID from the transaction results
        effects = result.result_data.effects
        created = effects.created
        
        if not created:
            raise Exception("No objects were created in the transaction")
            
        return created[0].reference.object_id
        
    def increment_counter(self, counter_id: str) -> None:
        """
        Increment a counter by 1.
        
        Args:
            counter_id: ID of the counter object
        """
        tx = MoveCallTransaction(
            self.client, 
            signer=self.config.active_address,
            package_object_id=self.package_id,
            module=self.counter_module,
            function="increment",
            arguments=[counter_id],
            type_arguments=[]
        )
        
        result = tx.execute()
        
        if not result.is_ok():
            raise Exception(f"Failed to increment counter: {result.result_string}")
    
    def increment_counter_by(self, counter_id: str, amount: int) -> None:
        """
        Increment a counter by a custom amount.
        
        Args:
            counter_id: ID of the counter object
            amount: Amount to increment by
        """
        tx = MoveCallTransaction(
            self.client, 
            signer=self.config.active_address,
            package_object_id=self.package_id,
            module=self.counter_module,
            function="increment_by",
            arguments=[counter_id, amount],
            type_arguments=[]
        )
        
        result = tx.execute()
        
        if not result.is_ok():
            raise Exception(f"Failed to increment counter: {result.result_string}")
    
    def reset_counter(self, counter_id: str) -> None:
        """
        Reset a counter to 0.
        
        Args:
            counter_id: ID of the counter object
        """
        tx = MoveCallTransaction(
            self.client, 
            signer=self.config.active_address,
            package_object_id=self.package_id,
            module=self.counter_module,
            function="reset",
            arguments=[counter_id],
            type_arguments=[]
        )
        
        result = tx.execute()
        
        if not result.is_ok():
            raise Exception(f"Failed to reset counter: {result.result_string}")
    
    def get_counter_value(self, counter_id: str) -> int:
        """
        Get the current value of a counter.
        
        Args:
            counter_id: ID of the counter object
            
        Returns:
            Current value of the counter
        """
        # Get the object data
        get_obj = GetObject(self.client, object_id=counter_id)
        result = get_obj.execute()
        
        if not result.is_ok():
            raise Exception(f"Failed to get counter object: {result.result_string}")
        
        # Extract the value field from the object data
        data = result.result_data
        fields = data.content.fields
        return int(fields.value)


def main():
    """
    Example usage of CounterClient.
    """
    import sys
    
    if len(sys.argv) < 2:
        print("Usage: python counter_client.py PACKAGE_ID [COUNTER_ID]")
        sys.exit(1)
    
    package_id = sys.argv[1]
    counter_id = sys.argv[2] if len(sys.argv) > 2 else None
    
    try:
        # Initialize the counter client
        client = CounterClient(package_id=package_id)
        
        # Create a new counter or use an existing one
        if not counter_id:
            print("Creating a new counter...")
            counter_id = client.create_counter()
            print(f"Created counter with ID: {counter_id}")
        
        # Get initial counter value
        value = client.get_counter_value(counter_id)
        print(f"Initial counter value: {value}")
        
        # Increment counter
        print("Incrementing counter...")
        client.increment_counter(counter_id)
        
        # Get updated value
        value = client.get_counter_value(counter_id)
        print(f"Counter value after increment: {value}")
        
        # Increment by custom amount
        increment_amount = 5
        print(f"Incrementing counter by {increment_amount}...")
        client.increment_counter_by(counter_id, increment_amount)
        
        # Get updated value
        value = client.get_counter_value(counter_id)
        print(f"Counter value after increment by {increment_amount}: {value}")
        
        # Reset counter
        print("Resetting counter...")
        client.reset_counter(counter_id)
        
        # Get updated value
        value = client.get_counter_value(counter_id)
        print(f"Counter value after reset: {value}")
        
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main() 

File: examples/typescript/counter_client.ts
import { TransactionBlock } from '@mysten/sui.js/transactions';
import { SuiClient } from '@mysten/sui.js/client';
import { Ed25519Keypair } from '@mysten/sui.js/keypairs/ed25519';
import { fromB64 } from '@mysten/sui.js/utils';

// Replace these with your values
const PRIVATE_KEY = 'YOUR_PRIVATE_KEY'; // Base64 encoded private key
const PACKAGE_ID = 'YOUR_PACKAGE_ID'; // Counter package ID
const COUNTER_MODULE = 'counter';
const COUNTER_OBJECT_ID = 'YOUR_COUNTER_OBJECT_ID'; // Optional: ID of an existing counter

class CounterClient {
    private client: SuiClient;
    private signer: Ed25519Keypair;
    private packageId: string;
    
    constructor(privateKey: string, packageId: string) {
        // Initialize Sui client
        this.client = new SuiClient({
            url: 'https://fullnode.mainnet.sui.io',
        });
        
        // Initialize signer
        const privateKeyBytes = fromB64(privateKey);
        this.signer = Ed25519Keypair.fromSecretKey(privateKeyBytes);
        
        // Set package ID
        this.packageId = packageId;
    }

    /**
     * Create a new counter object
     */
    async createCounter(): Promise<string> {
        const tx = new TransactionBlock();
        
        // Call the create_and_transfer function
        tx.moveCall({
            target: `${this.packageId}::${COUNTER_MODULE}::create_and_transfer`,
            arguments: [],
        });
        
        // Execute the transaction
        const result = await this.client.signAndExecuteTransactionBlock({
            signer: this.signer,
            transactionBlock: tx,
            options: {
                showEffects: true,
                showObjectChanges: true,
            },
        });
        
        // Find the created counter object ID from transaction effects
        const created = result.effects?.created;
        if (!created || created.length === 0) {
            throw new Error('Failed to create counter object');
        }
        
        // Return the ID of the created counter
        return created[0].reference.objectId;
    }

    /**
     * Increment a counter by 1
     * @param counterId The counter object ID
     */
    async incrementCounter(counterId: string): Promise<void> {
        const tx = new TransactionBlock();
        
        // Call the increment function
        tx.moveCall({
            target: `${this.packageId}::${COUNTER_MODULE}::increment`,
            arguments: [tx.object(counterId)],
        });
        
        // Execute the transaction
        await this.client.signAndExecuteTransactionBlock({
            signer: this.signer,
            transactionBlock: tx,
        });
    }

    /**
     * Increment a counter by a custom amount
     * @param counterId The counter object ID
     * @param amount The amount to increment by
     */
    async incrementCounterBy(counterId: string, amount: number): Promise<void> {
        const tx = new TransactionBlock();
        
        // Call the increment_by function
        tx.moveCall({
            target: `${this.packageId}::${COUNTER_MODULE}::increment_by`,
            arguments: [
                tx.object(counterId),
                tx.pure(amount)
            ],
        });
        
        // Execute the transaction
        await this.client.signAndExecuteTransactionBlock({
            signer: this.signer,
            transactionBlock: tx,
        });
    }

    /**
     * Reset a counter to 0
     * @param counterId The counter object ID
     */
    async resetCounter(counterId: string): Promise<void> {
        const tx = new TransactionBlock();
        
        // Call the reset function
        tx.moveCall({
            target: `${this.packageId}::${COUNTER_MODULE}::reset`,
            arguments: [tx.object(counterId)],
        });
        
        // Execute the transaction
        await this.client.signAndExecuteTransactionBlock({
            signer: this.signer,
            transactionBlock: tx,
        });
    }

    /**
     * Get the counter value
     * @param counterId The counter object ID
     */
    async getCounterValue(counterId: string): Promise<number> {
        // Get object data
        const object = await this.client.getObject({
            id: counterId,
            options: {
                showContent: true,
            },
        });
        
        // Extract the value field
        if (
            object &&
            object.data &&
            object.data.content &&
            'fields' in object.data.content
        ) {
            return Number(object.data.content.fields.value);
        }
        
        throw new Error('Failed to get counter value');
    }
}

// Example usage
async function main() {
    try {
        const counterClient = new CounterClient(PRIVATE_KEY, PACKAGE_ID);
        
        // Create a new counter or use an existing one
        let counterId = COUNTER_OBJECT_ID;
        if (!counterId) {
            console.log('Creating a new counter...');
            counterId = await counterClient.createCounter();
            console.log(`Created counter with ID: ${counterId}`);
        }
        
        // Get initial counter value
        let value = await counterClient.getCounterValue(counterId);
        console.log(`Initial counter value: ${value}`);
        
        // Increment counter
        console.log('Incrementing counter...');
        await counterClient.incrementCounter(counterId);
        
        // Get updated value
        value = await counterClient.getCounterValue(counterId);
        console.log(`Counter value after increment: ${value}`);
        
        // Increment by custom amount
        const incrementAmount = 5;
        console.log(`Incrementing counter by ${incrementAmount}...`);
        await counterClient.incrementCounterBy(counterId, incrementAmount);
        
        // Get updated value
        value = await counterClient.getCounterValue(counterId);
        console.log(`Counter value after increment by ${incrementAmount}: ${value}`);
        
        // Reset counter
        console.log('Resetting counter...');
        await counterClient.resetCounter(counterId);
        
        // Get updated value
        value = await counterClient.getCounterValue(counterId);
        console.log(`Counter value after reset: ${value}`);
        
    } catch (error) {
        console.error('Error:', error);
    }
}

// Uncomment to run the example
// main();

export { CounterClient }; 